File: samples/python/visual_language_chat/prompt_lookup_decoding_vlm_chat.py
#!/usr/bin/env python3
# Copyright (C) 2026 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import numpy as np
from PIL import Image
from pathlib import Path

import openvino_genai
from openvino import Tensor


def streamer(subword):
    print(subword, end="", flush=True)
    # Return flag corresponds whether generation should be stopped.
    return openvino_genai.StreamingStatus.RUNNING


def read_image(path: str) -> Tensor:
    """

    Args:
        path: The path to the image.

    Returns: the ov.Tensor containing the image.

    """
    pic = Image.open(path).convert("RGB")
    image_data = np.array(pic)
    return Tensor(image_data)


def read_images(path: str) -> list[Tensor]:
    entry = Path(path)
    if entry.is_dir():
        return [read_image(str(file)) for file in sorted(entry.iterdir())]
    return [read_image(path)]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("model_dir")
    parser.add_argument("image_dir", help="Image file or dir with images")
    parser.add_argument(
        "--disable_lookup",
        action="store_false",
        dest="enable_lookup",
        default=True,
        help="Disable lookup decoding (i.e., set enable_lookup to False).",
    )

    args = parser.parse_args()

    device = "CPU"

    # Prompt lookup decoding in VLM pipeline enforces ContinuousBatching backend.
    pipe = openvino_genai.VLMPipeline(args.model_dir, device, prompt_lookup=args.enable_lookup)

    config = openvino_genai.GenerationConfig()
    config.max_new_tokens = 100
    if args.enable_lookup:
        # add parameter to enable prompt lookup decoding to generate `num_assistant_tokens` candidates per iteration
        config.num_assistant_tokens = 5
        # Define max_ngram_size
        config.max_ngram_size = 3

    rgbs = read_images(args.image_dir)

    chat_history = openvino_genai.ChatHistory()
    while True:
        try:
            prompt = input("\n----------\nquestion:\n")
        except EOFError:
            break
        # New images and videos can be passed at each turn
        chat_history.append({"role": "user", "content": prompt})
        decoded_results: openvino_genai.DecodedResults = pipe.generate(
            chat_history, images=rgbs, generation_config=config, streamer=streamer
        )
        chat_history.append({"role": "assistant", "content": decoded_results.texts[0]})
    print()


if __name__ == "__main__":
    main()

File: samples/cpp/visual_language_chat/prompt_lookup_decoding_vlm_chat.cpp
// Copyright (C) 2023-2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include <openvino/openvino.hpp>

#include "load_image.hpp"
#include "openvino/genai/visual_language/pipeline.hpp"

int main(int argc, char* argv[]) try {
    if (3 != argc) {
        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> <IMAGE_FILE OR DIR_WITH_IMAGES>");
    }

    ov::genai::GenerationConfig config;
    config.max_new_tokens = 100;
    // Define candidates number for candidate generation
    config.num_assistant_tokens = 5;
    // Define max_ngram_size
    config.max_ngram_size = 3;

    std::string model_path = argv[1];
    std::vector<ov::Tensor> rgbs = utils::load_images(argv[2]);
    std::string prompt;
    std::string device = "CPU";

    // Prompt lookup decoding in VLM pipeline enforces ContinuousBatching backend.
    ov::genai::VLMPipeline pipe(
        model_path,
        device,
        ov::genai::prompt_lookup(true));

    auto streamer = [](std::string subword) {
        std::cout << subword << std::flush;
        return ov::genai::StreamingStatus::RUNNING;
    };

    // Since the streamer is set, the results are
    // printed each time a new token is generated.
    ov::genai::ChatHistory chat_history;

    std::cout << "\n----------\nquestion:\n";
    while (std::getline(std::cin, prompt)) {
        chat_history.push_back({{"role", "user"}, {"content", prompt}});
        auto decoded_results = pipe.generate(chat_history,
                                             ov::genai::images(rgbs),
                                             ov::genai::generation_config(config),
                                             ov::genai::streamer(streamer));
        std::string output = decoded_results.texts[0];
        chat_history.push_back({{"role", "assistant"}, {"content", output}});
        std::cout << "\n----------\nquestion:\n";
    }
} catch (const std::exception& error) {
    try {
        std::cerr << error.what() << '\n';
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
} catch (...) {
    try {
        std::cerr << "Non-exception object thrown\n";
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
}
